packages/discovery-provider/plugins/pedalboard/apps/backfill-audio-analyses/src/backfill.ts

```diff
@@ -0,0 +1,164 @@
+        `${analysisUrl}?${queryParams.toString()}`,
+        null,
+        { timeout: REQUEST_TIMEOUT }
+      )
+      if (response.status !== 200) {
+        console.log(
+          `Received ${response.status} response from ${contentNodeEndpoint}. Trying another content node...`
+        )
+        continue
+      }
+    } catch (error: any) {
+      if (error.isAxiosError !== undefined && error.code === 'ECONNABORTED') {
+        console.log(
+          `Timeout error triggering audio analysis for track ID ${track.track_id} on ${contentNodeEndpoint}: ${error.message}. Trying another content node...`
+        )
+        continue
+      } else {
+        console.error(
+          `Error analyzing audio for track ID ${track.track_id}, upload ID ${
+            track.audio_upload_id
+          }: ${(error as Error).message}. Skipping track...`
+        )
+        break
+      }
+    }
+
+    release() // Release the semaphore permit
+  }
+}
+
+async function fetchTracks(
+  offset: number,
+  limit: number,
+  db: Knex
+): Promise<Track[]> {
+  return await db<Track>('tracks')
+    .where(function () {
+      this.whereNull('musical_key').orWhereNull('bpm')
+    })
+    .andWhere('audio_analysis_error_count', 0)
+    .orderBy('track_id', 'asc')
+    .offset(offset)
+    .limit(limit)
+}
+
+// Trigger audio analyses for BACKFILL_BATCH_SIZE tracks at a time
+async function processBatches(db: Knex): Promise<void> {
+  let offset
+  while (true) {
+    console.time('Batch processing time')
+    const contentNodes = await getCachedHealthyContentNodes()
+    if (contentNodes.length == 0) {
+      console.timeEnd('Batch processing time')
+      console.error(`No healthy content nodes found. Please investigate`)
+      return
+    }
+    // Take 5 random endpoints to query for audio analysis
+    const endpoints = _.shuffle(
+      contentNodes.map((node) => node.endpoint)
+    ).slice(0, 5)
+    offset = await readDbOffset()
+    if (offset == null) {
+      offset = 0
+    }
+    const tracks = await fetchTracks(offset, BACKFILL_BATCH_SIZE, db)
+    if (tracks.length === 0) {
+      console.timeEnd('Batch processing time')
+      break
+    }
+    const analyzePromises = tracks.map((track) =>
+      analyzeAudio(endpoints, track)
+    )
+    await Promise.all(analyzePromises)
+
+    offset += BACKFILL_BATCH_SIZE
+    await storeDbOffset(offset)
+    console.timeEnd('Batch processing time')
+    console.log(
+      `Triggered audio analyses for ${tracks.length} tracks. New offset: ${offset}`
+    )
+
+    if (config.testRun) {
+      console.log(
+        `[TEST RUN] Triggered audio analyses for the following track IDs: ${tracks.map(
+          (track) => track.track_id
+        )}`
+      )
+      // only do 1 batch in a test run
+      break
+    }
+
+    // Sleep for 30 seconds
+    await new Promise((resolve) => setTimeout(resolve, 30000))
+  }
+
+  console.log('No more tracks to backfill. Sleeping forever...')
+  await new Promise(() => {})
```

Comment:
this will hold up memory won't it? do we want to exit with a 0 status?

Reply:
could also build in the repair task here that'll live forever

Reply:
tru, i made it sleep so it compose wouldn't auto restart it but exiting with 0 status and a restart `on-failure` policy should be fine. will add

thought about putting the repairer here but its a little conceptually closer to the other discovery tasks / more findable as a celery task